brainstorming.md

```diff
@@ -0,0 +1,372 @@
+
+- support for accessing data from persistent, cloud-resident datasets,
+- enable use within popular data science environments (eg, R, Python), using the tools of that environment (eg, Python Pandas integration),
+- enable access to data aggregations much larger than single-host main memory
+- provide a building block for higher-level API that may embody domain-specific conventions or schema around annotated 2D matrices (eg, a cell "atlas").
+
+The SOMA data model is centered on annotated 2-D matrices, conceptually similar to data structures embedded within existing single-cell data models including Seurat Assay, Bioconductor SingleCellExperiment, and ScanPy AnnData. Where possible, the SOMA API attempts to be general purpose and agnostic to the specifics of any given environment, or to the specific conventions of the Single Cell scientific ecosystem.
+
+SOMA is an abstract _API specification_, with the goal of enabling multiple concrete API implementations within different computing environments and data storage systems. SOMA does not specify an at-rest serialization format or underlying storage system.
+
+This document is attempting to codify the abstract, language-neutral SOMA data model and functional operations. Other specifications will document specific language bindings and particular storage system implementations.
+
+# Data Model
+
+The abstract data model comprises three "core" data types, and one specialization intended to improve ease of use and interoperability. The core types are:
+
+- SOMACollection - a string-keyed container (key-value map) of other SOMA data types, eg, SOMADataFrame, SOMAMatrix and SOMACollection.
+- SOMADataFrame - a 1-D, string-indexed, multi-column table, with monomorphic columns of equal length -- essentially a single-index dataframe.
+- SOMAMatrix - a 2-D, sparse, string-indexed multi-column matrix, with monomorphic columns of equal length -- essentially a two-index dataframe.
+
+In addition, a convenience type is defined:
+
+- SOMA - a specialization and extension of SOMACollection, codifying a set of naming and indexing conventions to represent an annotated, 2-D matrix of observations across a well-defined set of variables.
+
+In this document, the term `dataframe` implies something akin to an Arrow `RecordBatch`, R `data.table` or Python `pandas.DataFrame`, where:
+
+- dimensions are string-indexed
+- multiple columns may exist, each with a string column name
+- all columns are individually typed, monomorphic, and contain simple types (eg, int64)
+- all columns are of equal length
+
+All SOMA data objects are named with URIs.
+
+> ⚠️ **Issue** -- this section would benefit from a data model diagram, helpfing visualize the relationships between the types.
+
+> ⚠️ **Issue** -- the use of `SOMA` to name the API _and_ a data type is confusing. We should consider a new name for the aforementioned SOMA _type_. _Candidate:_ `SOMAAnnotatedMatrix`?
```

Comment:
really good call -- this is another 'high-level issue' we need to deal with to avoid uptake friction

i like:

* keep SOMA API for the API
* keep for the (current) object, rename it away from `SOMA` to something else -- not sure what -- open to ideas

Reply:
or `SOMAAnnotationMatrixEnsemble`?

Reply:
@ambrosejcarr - would love your thoughts on this.